### PR TITLE
Preprocessor to resolve ifdefs is applied to shader programs

### PR DESCRIPTION
**Pull request by mvaligursky**
A new private class is added to preprocess defines and ifdefs in the shader with the goal of removing unused code. This is a first preparation step. Following steps will further modify the shader code, in order to convert uniforms to uniform blocks.

**New (private) API:**
```
/**
 * Run c-like proprocessor on the source code, and resolves the code based on the defines and ifdefs
 *
 * @param {string} source - The source code to work on.
 * @returns {string|null} Returns preprocessed source code.
 */
static Preprocessor.run(source) {
```

**Note** - Simple expression evaluation only is supported for ifdefs, handles cases:
- expression
- defined(expression)
- !defined(expression)
 
 It does not handle more complex cases, which would require more complex system:
 - defined(A) || defined(B)
And so few use cases in the shaders were converted to the supported format.

Unit test has also been added.

Additionally, Debug logging functionality has been extended with **trace** function.
**New (private) API (and also examples of use):**
```
Debug.setTrace(`Preprocessor`, true);
Debug.trace(`Preprocessor`, "message")
```

and this is displayed in the console:
`[Preprocessor]  message`

### CODE REVIEW

**Comment by yaustar** [comment]
Does this affect the sourcemap generation?

**Comment by mvaligursky** [comment]
> Does this affect the sourcemap generation?

This is only used to process shaders in memory .. so no.

**Comment by willeastcott** [comment]
This is definitely something for @slimbuck to review - I'll pass.

File: src/core/preprocessor.js [hunk 1]
```diff
@@ -0,0 +1,256 @@
+import { Debug } from './debug.js';
+
+// set to true to enable debug logging
+const enableLogging = false;
```
Comment (slimbuck) [inline src/core/preprocessor.js:4]: I love this idea, really useful. It does mean the log calls in this file remain in production code though.

Perhaps we could embrace a more general approach to trace logging?

We could add `Debug.trace` function which takes a named "stream" parameter and allow user to enable/disable streams at runtime. Calls to `Debug.trace` (and the stream enable calls) would be removed by rollup as usual.

File: src/core/preprocessor.js [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+ * @ignore
+ */
+class Preprocessor {
+    /**
+     * Run c-like proprocessor on the source code, and resolves the code based on the defines and ifdefs
+     *
+     * @param {string} source - The source code to work on.
+     * @returns {string} Returns preprocessed source code.
+     */
+    static run(source) {
```
Comment (slimbuck) [inline src/core/preprocessor.js:53]: Shouldn't this function return success or failure?

File: src/core/preprocessor.js (hunk 1 above)
Comment (mvaligursky) [reply src/core/preprocessor.js]: Yep I thought about it few times as well .. and now is a good time to do it. I'll add description to this PR.

File: src/graphics/program-lib/chunks/lit/clusteredLight.frag.js [hunk 3]
```diff
@@ -3,6 +3,17 @@ uniform sampler2D clusterWorldTexture;
 uniform sampler2D lightsTexture8;
 uniform highp sampler2D lightsTextureFloat;
 
+// complex ifdef expression are not supported, handle it here
+// defined(CLUSTER_COOKIES) || defined(CLUSTER_SHADOWS)
+#if defined(CLUSTER_COOKIES)
+    #define CLUSTER_COOKIES_OR_SHADOWOWS
+#endif
+#if defined(CLUSTER_SHADOWS)
+    #ifndef CLUSTER_COOKIES_OR_SHADOWOWS
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/graphics/program-lib/chunks/lit/clusteredLight.frag.js]: I assumed those are needed .. but clearly not. Removed, thanks.

File: src/core/preprocessor.js (hunk 2 above)
Comment (mvaligursky) [reply src/core/preprocessor.js:43]: it's little error validation taking place in this function, but I return error (null) and handle it on the other side correctly now (there was an unrelated bug in shader compilation returning incorrect type)

**Comment by mvaligursky** [comment]
Updates based on comments @slimbuck 
@willeastcott - could you please look at the Debug.trace part I added.

File: rollup.config.js [hunk 4]
```diff
@@ -136,6 +136,7 @@ const stripOptions = {
         'Debug.warn',
         'Debug.error',
         'Debug.log',
+        'Debug.trace',
```
Comment (slimbuck) [inline rollup.config.js:139]: Is possible (/does it make sense) to remove `setTrace` and `_traceChannels` too?

**Review by slimbuck — APPROVED** [review]
We'll probably update the preprocessor over time, but this is a reasonable first pass.

File: rollup.config.js (hunk 4 above)
Comment (mvaligursky) [reply rollup.config.js:139]: _traceChannels is automatically stripped (the same as the existing other set).
setTrace is an empty function in non debug build which I'd like to keep in .. allowing the code to compile in non-debug builds as well.

Comment (slimbuck) [reply rollup.config.js:139]: Wouldn't the rollup plugin remove references in the code to 'Debug.setTrace' as well as the function itself?

Comment (slimbuck) [reply rollup.config.js:139]: Isn't that an issue for `Debug.trace` anyway?

Comment (mvaligursky) [reply rollup.config.js:139]: well, Debug.trace is to be used inside the engine only .. but Debug.setTrace is called from outside to enabled the tracing. That's my plan on how this should be used.

Comment (slimbuck) [reply rollup.config.js:139]: I see, makes sense 👍

**Pull request merged**